controlplane/kubeadm/controllers/kubeadm_control_plane_controller.go

```diff
@@ -331,18 +334,71 @@ func (r *KubeadmControlPlaneReconciler) updateStatus(ctx context.Context, kcp *c
 	return nil
 }
 
-func (r *KubeadmControlPlaneReconciler) upgradeControlPlane(ctx context.Context, cluster *clusterv1.Cluster, kcp *controlplanev1.KubeadmControlPlane, requireUpgrade []clusterv1.Machine) error {
-
-	// TODO: verify health for each existing replica
-	// TODO: mark an old Machine via the label kubeadm.controlplane.cluster.x-k8s.io/selected-for-upgrade
-	// TODO: check full cluster health
-	// TODO: provision new Machine to replace marked Machine
-	// TODO: wait for health of all existing replicas, ideally in a re-entrant way to avoid waiting in process
-	// TODO: wait for full cluster health, ideally in a re-entrant way to avoid waiting in process
-	// TODO: Remove etcd membership for old Machine instance
-	// TODO: Update the kubeadm configmap
-	// TODO: Delete the Marked ControlPlane machine
-	// TODO: Continue with next OldMachine
+func (r *KubeadmControlPlaneReconciler) upgradeControlPlane(ctx context.Context, cluster *clusterv1.Cluster, kcp *controlplanev1.KubeadmControlPlane, logger logr.Logger) error {
+	// TODO: handle reconciliation of etcd members and kubeadm config in case they get out of sync with cluster
+
+	ownedMachines, err := r.managementCluster.GetMachinesForCluster(ctx, clusterKey(cluster), internal.OwnedControlPlaneMachines(kcp.Name))
+	if err != nil {
+		return err
+	}
+	requireUpgrade := internal.FilterMachines(
+		ownedMachines,
+		internal.HasOutdatedConfiguration(hash.Compute(&kcp.Spec)),
+		internal.OlderThan(kcp.Spec.UpgradeAfter),
+	)
+
+	if err := r.managementCluster.TargetClusterEtcdIsHealthy(ctx, clusterKey(cluster), kcp.Name); err != nil {
+		logger.Error(err, "Waiting for etcd cluster to pass health check")
+		r.recorder.Eventf(kcp, corev1.EventTypeWarning, "ETCDClusterUnhealthy", "Waiting for etcd cluster to be healthy to continue upgrade: %v", err)
+		return &capierrors.RequeueAfterError{RequeueAfter: 30 * time.Second}
+	}
+	if err := r.managementCluster.TargetClusterControlPlaneIsHealthy(ctx, clusterKey(cluster), kcp.Name); err != nil {
+		logger.Error(err, "Waiting for control plane to pass health check")
+		r.recorder.Eventf(kcp, corev1.EventTypeWarning, "ControlPlaneUnhealthy", "Waiting for control plane to be healthy to continue upgrade: %v", err)
+		return &capierrors.RequeueAfterError{RequeueAfter: 30 * time.Second}
+	}
+
+	// If there is not already a Machine that is marked for upgrade, find one and mark it
+	selectedForUpgrade := internal.FilterMachines(requireUpgrade, internal.SelectedForUpgrade())
+	if len(selectedForUpgrade) == 0 {
+		sort.Slice(requireUpgrade, func(i, j int) bool {
+			return requireUpgrade[i].CreationTimestamp.Before(&requireUpgrade[j].CreationTimestamp)
+		})
+
+		oldest := &requireUpgrade[0]
```

Comment:
this could, in theory, hit an out of index. Since this is getting a new `requireUpgrade` list of machines inside the function and not checking for length > 0, it's possible, however unlikely there could be no machines at this point that require upgrade.

The real question I have is: what was the reason for changing this from passing in a list of requireUpgrade machines to recalculate the list of requireUpgrade machines?

Reply:
The main impetus was that we need access to the full list of `ownedMachines` and it seemed a bit odd to pass both in when we could easily recompute. I'm fine going with passing in both slices if you think it makes more sense.

Reply:
ah gotcha, nah recompute is fine, just wanted to understand why the change. I don't think this is a likely edge case anyway.  thanks for the insight